File: Elixir/lib/command.ex
defmodule ServerBench.Command do
  @enforce_keys [:type]
  defstruct [:type, key: nil, value: nil, duration: nil]
  def string_valid?(input) do
    regexp = ~r/[a-zA-Z0-9]+/
    Regex.match?(regexp, input)
  end

  def parse(input) when is_bitstring(input) do
    String.trim(input)
    |> String.split()
    |> parse()
  end

  def parse(["GET", key]) do
    if string_valid?(key) do
      %ServerBench.Command{type: :get, key: key}
    else
      %ServerBench.Command{type: :invalid}
    end
  end

  def parse(["SET", key, value]) do
    if string_valid?(key) && string_valid?(value) do
      %ServerBench.Command{type: :set, key: key, value: value}
    else
      %ServerBench.Command{type: :invalid}
    end
  end

  def parse(["DEL", key]) do
    if string_valid?(key) do
      %ServerBench.Command{type: :del, key: key}
    else
      %ServerBench.Command{type: :invalid}
    end
  end

  def parse(["GETC"]) do
    %ServerBench.Command{type: :get_counter}
  end

  def parse(["SETC"]) do
    %ServerBench.Command{type: :set_counter}
  end

  def parse(["DELC"]) do
    %ServerBench.Command{type: :del_counter}
  end

  def parse(["GETDUMP"]) do
    %ServerBench.Command{type: :get_dump}
  end

  def parse(["NEWDUMP"]) do
    %ServerBench.Command{type: :new_dump}
  end

  def parse(["DUMPINTERVAL", timestamp]) when is_bitstring(timestamp) do
    parse(["DUMPINTERVAL", parse_time(timestamp)])
  end

  def parse(["DUMPINTERVAL", nil]) do
    %ServerBench.Command{type: :invalid}
  end

  def parse(["DUMPINTERVAL", [_, hours, minutes, seconds]]) do
    time = calc_time(String.to_integer(hours), String.to_integer(minutes), String.to_integer(seconds))
    %ServerBench.Command{type: :dump_interval, duration: time}
  end

  def parse(["SETTTL", key, value, timestamp]) when is_bitstring(timestamp) do
    if string_valid?(key) && string_valid?(value) do
      parse(["SETTTL", key, value, parse_time(timestamp)])
    else

      %ServerBench.Command{type: :invalid}
    end
  end

  def parse(["SETTTL", _, _, nil]) do
    %ServerBench.Command{type: :invalid}
  end

  def parse(["SETTTL", key, value, [_, hours, minutes, seconds]]) do
    %ServerBench.Command{type: :set_ttl, key: key, value: value, duration: calc_time(String.to_integer(hours), String.to_integer(minutes), String.to_integer(seconds))}
  end

  def parse(_) do
    %ServerBench.Command{type: :invalid}
  end

  def parse_time(input) when is_bitstring(input) do
    time_regex = ~r/([0-9][0-9])h-([0-9][0-9])m-([0-9][0-9])s/
    matches = Regex.run(time_regex, String.trim(input))
    matches
  end

  def parse_time(_) do
    IO.puts "Time parse failed"
  end

  def calc_time(hours, minutes, seconds) do
    ret = seconds
    ret = ret + (minutes * 60)
    ret + (hours * 60 * 60)
  end
end

File: Ruby/command.rb
GET_STRING = 'GET'
SET_STRING = 'SET'
DEL_STRING = 'DEL'
GET_COUNTER_STRING = 'GETC'
SET_COUNTER_STRING = 'SETC'
DEL_COUNTER_STRING = 'DELC'
GET_DUMP_STRING = 'GETDUMP'
NEW_DUMP_STRING = 'NEWDUMP'
DUMP_INTERVAL_STRING = 'DUMPINTERVAL'
SET_TTL_STRING = 'SETTTL'

VALUE_REGEX = Regexp.new "[a-zA-Z0-9]+"
TIME_REGEX = Regexp.new "([0-9][0-9])h-([0-9][0-9])m-([0-9][0-9])s"

module Parsing
  def parse_type(part)
    case part
    when GET_STRING
      return :get
    when SET_STRING
      return :set
    when DEL_STRING
      return :del
    when GET_COUNTER_STRING
      return :get_counter
    when SET_COUNTER_STRING
      return :set_counter
    when DEL_COUNTER_STRING
      return :del_counter
    when GET_DUMP_STRING
      return :get_dump
    when NEW_DUMP_STRING
      return :new_dump
    when DUMP_INTERVAL_STRING
      return :dump_interval
    when SET_TTL_STRING
      return :set_ttl
    else
      return :invalid
    end
  end

  def validate_string(input)
    VALUE_REGEX.match(input) != nil
  end

  def parse_time(input)
    match_result = TIME_REGEX.match input
    if match_result == nil
      return nil
    end

    hours_string = match_result[0]
    minutes_string = match_result[1]
    seconds_string = match_result[2]
    hours = hours_string.to_i
    minutes = minutes_string.to_i
    seconds = seconds_string.to_i
    seconds += minutes * 60
    seconds += hours * 60 * 60
    seconds
  end

  def parse_command(line)
    if line == nil
      command = Command.new
      command.type = :invalid
      return command
    end
    parts = line.split ' '
    if parts.length < 1
      return nil
    end

    command = Command.new

    type_string = parts[0]

    command.type = parse_type type_string

    if command.type == :invalid
      return nil
    end


    case command.type
    when :get
      if parts.length != 2
        return nil
      end
      key_string = parts[1]
      unless validate_string(key_string)
        return nil
      end
      command.key = key_string
    when :set
      if parts.length != 3
        return nil
      end
      key_string = parts[1]
      unless validate_string(key_string)
        return nil
      end
      command.key = key_string
      value_string = parts[2]
      unless validate_string value_string
        return nil
      end
      command.value = value_string
    when :del
      if parts.length != 2
        return nil
      end
      key_string = parts[1]
      unless validate_string key_string
        return nil
      end
      command.key = key_string
    when :get_counter
      if parts.length != 1
        return nil
      end
    when :set_counter
      if parts.length != 1
        return nil
      end
    when :del_counter
      if parts.length != 1
        return nil
      end
    when :get_dump
      if parts.length != 1
        return nil
      end
    when :new_dump
      if parts.length != 1
        return nil
      end
    when :dump_interval
      if parts.length != 2
        return nil
      end

      time = parse_time parts[1]
      if time == nil
        return nil
      end
      command.time = time
    when :set_ttl
      if parts.length != 4
        return nil
      end
      key_string = parts[1]
      unless validate_string key_string
        return nil
      end
      value_string = parts[2]
      unless validate_string value_string
        return nil
      end
      time_string = parts[3]
      time = parse_time time_string
      if time == nil
        return nil
      end
      command.key = key_string
      command.value = value_string
      command.time = time
    else
      return nil
    end

    command
  end
end



class Command
  @type
  @key
  @value
  @time

  def initialize(type = :invalid, key = nil, value = nil, time = nil)
    @type = type
    @key = key
    @value = value
    @time = time
  end

  attr_accessor :type, :key, :value, :time

  def to_s
    "#{@type} #{@key} #{@value} #{@time}"
  end
end
